docs/content/sips/017-service-chaining.md

```diff
@@ -0,0 +1,122 @@
+
+# User experience
+
+## Expressing a requirement for chaining
+
+To express that the application requires service chaining, add the following to `spin.toml`:
+
+```toml
+[application.host_requirements]
+spin_internal_requests = "required"
+```
+
+> Why ask the user to write out `"required"` instead of just listing the names?  In future, we may have features where it makes sense to say "I'd prefer to run on a host that supports this, but it's not a blocker," or where the user can specify a value (for a bad example, consider an "available memory" prerequisite).
+
+If a host doesn't understand the `host_requirements` section, it must reject the application.  We can't retro-fit this to existing hosts so we have to rely on `serde` for this, which means the error will be unlovely.  However, for individual fields within the section, hosts can give better errors/warnings.
+
+If a host understands the section, but doesn't understand an entry in that section, it _must_ reject the application with an error message explaining the required host features.
+
+## Requesting a component via chaining
+
+There are two ways we could express a HTTP request to a chained component:
+
+1. Use the existing self-request style.  The host _may_ chain self-requests, and _must_ chain them if the "require chaining" setting is on.  This is super easy for the application developer, at both the send and receive end.  However, it becomes more complex for the host, as the host now has to route the request - this drags in much more of the HTTP trigger plumbing.  (The prototype did this and it was egregiously hairy.)
+
+2. Address requests directly to the destination components by name. We still need a URL, so this would introduce a special host name such as `spin.internal`.  Senders address individual components via subdomains.  Thus, the URL `http://accounts.spin.internal` would invoke the `accounts` component.  This bypasses routing, which should simplify implementation.  It also opens the door to private endpoints, where an internal microservice is not exposed to the network (does not have a HTTP route) and so is accessible only from within the application.  However, chained components which parse URLs may need to take extra care to correctly handle both routed and chaining URLs.
+
+**The preferred user experience is option 2: `http://<component-name>.spin.internal/`**
+
+> This is why we cannot have "preferred but not required" for internal requests.  Using the `foo.spin.internal` host requires direct component addressing by the host - we cannot fall back to making a routed request.
+
+## Incoming request headers
+
+To avoid spoofing, Spin must strip any `Host: *.spin.internal` headers from routed requests.
```

Comment:
Decision: agreed to block this header pattern, and to ensure that chained requests supply the header.

Reply:
If sender sets Host header to something else then we overwrite it.